### DISCUSSION TITLE
How are duplicate values in IncrementalValuesProvider handled?

### DISCUSSION THREAD

**Discussion by rick-de-water** [dotnet/roslyn#80820] [Q&A]
Lets say I am generating code for some properties in this class.
```csharp
public partial class Foo
{
    [MyAttribute] public int X { get; set; }
    [MyAttribute] public int Y { get; set; }
}
```

And I use the syntax provider to look for fields with this attibute, and when one is discovered it returns the parent class.
```csharp
IncrementalValuesProvider<ClassDeclarationSyntax> classes = context.SyntaxProvider
    .CreateSyntaxProvider(
        static (node, _) => { /* Return PropertyDeclaractionSyntax with at least 1 attribute */ },
        static (context, _) => { /* If property has MyAttribute return node.Parent */ })
    .Where(static node => node != null);
```

Now the `classes` value provider contains the same class twice, since both properties are part of the same class. What happens here in terms of caching and source output? Is the first value processed and will anything that uses the second value use the cache? Would it generate a source output once or twice if this was hooked up? If that is the case, is there a way to apply something like `Distinct` to value providers?

I seem to run into this scenario quite often, since I want to listen to properties and fields with attributes, but then I want to generate code for the entire class. I usually end up using `Collect` somewhere, but I would like to avoid that since then a single change can cause a big part of the pipeline to run again.

**Comment by AraHaan** [comment]
After some debugging, it seems I run across a similar issue where I filter between ``ClassDeclarationSyntax``'s.

```
Work Item 0: Namespace: Name: MyApp
Class Item 0: Class: Name TestDisposable, Accessibility: Internal, Stream: False, Owns Count: 0, Fields Count: 0, SetNull Count: 1, Methods Count: 0
Work Item 0: Namespace: Name: MyApp.AnotherNamespace
Class Item 0: Class: Name AnotherDisposable, Accessibility: Internal, Stream: False, Owns Count: 0, Fields Count: 0, SetNull Count: 0, Methods Count: 0
```
In my case I expected my debug to print:

```
Work Item 0: Namespace: Name: MyApp
Class Item 0: Class: Name TestDisposable, Accessibility: Internal, Stream: False, Owns Count: 0, Fields Count: 0, SetNull Count: 1, Methods Count: 0
Work Item 1: Namespace: Name: MyApp.AnotherNamespace
Class Item 0: Class: Name AnotherDisposable, Accessibility: Internal, Stream: False, Owns Count: 0, Fields Count: 0, SetNull Count: 0, Methods Count: 0
```
Where it would use only a single instance of my own collection type per ``Compilation`` but instead it created 2 on this test which made it fail.

<details>
<summary>Issue:</summary>

Basically in this case, I cant sadly have ``CreateSyntaxProvider`` group the found SyntaxNodes based on my compilation to where I can simply have a single instance of my collection. I cannot new it up in ``context.CompilationProvider`` due to also needing to new up in ``CreateSyntaxProvider`` because I cannot pass it into it (to reuse it). This results in an issue that I cannot resolve easily at all.

Basically what I (and the person who filed this issue) needs is:

```cs
/// <summary>
/// Creates an <see cref="IncrementalValueProvider{TResult}"/> that can provide a transform over <see cref="SyntaxNode"/>s
/// </summary>
/// <typeparam name="TSource">The type of the value of the consumed <see cref="IncrementalValueProvider{TSource}"/></typeparam>
/// <typeparam name="TResult">The type of the value the syntax node is transformed into</typeparam>
/// <param name="predicate">A function that determines if the given <see cref="SyntaxNode"/> should be transformed</param>
/// <param name="transform">A function that performs the transform, when <paramref name="predicate"/>returns <c>true</c> for a given node</param>
/// <returns>An <see cref="IncrementalValueProvider{TResult}"/> that provides the results of the transformation</returns>
public IncrementalValuesProvider<TResult> CreateSyntaxProvider<TSource, TResult>(IncrementalValueProvider<TSource> source, Func<SyntaxNode, CancellationToken, bool> predicate, Func<GeneratorSyntaxContext, CancellationToken, TSource, TResult> transform);
```

Where we can simply ignore the results, and be able to access the processed data through the data passed in from the ``source`` parameter.

Sadly I would provide my own implementation for it as an extension method, however the ``IncrementalValuesProvider``, and the ``SyntaxInputNode`` (actually it might need another constructor as well so it can pass in the value from the source parameter) constructors are not exposed publicly at all (at least in the roslyn 4.0.1 packages). Besides the ``WrapUserFunction`` would need a new overload for this as well too that takes ``TInput1, TInput2, TOutput``.
</details>

<details>
<summary>Workaround:</summary>

I could only work around this my making my own dummy collection type accept in a ``Compilation`` in it's contructor to bind it's instance to that of the compilation.

```cs
var workItemCollection = context.CompilationProvider.Select(
    static (c, _) => new WorkItemCollection(c));
var workItems = context.SyntaxProvider.CreateSyntaxProvider(
    static (n, _) => n is ClassDeclarationSyntax,
    (n, ct) => (INamedTypeSymbol)n.SemanticModel.GetDeclaredSymbol(n.Node, ct)!
    ).Combine(workItemCollection).Select(
    static (testClass, ct) =>
    {
        testClass.Right.Process(testClass.Left, ct);
        return true;
    });
var combined = workItems.Collect().Combine(workItemCollection);
```

After that, when I use ``combined`` in ``context.RegisterSourceOutput`` all 23 of my tests started to pass and the generator seems to have started working as intended.
</details>

**Comment by AraHaan** [comment]
@rick-de-water what you could do instead is filter by ClassDeclarationSyntax then to the processing of it to find the attribute on the members. If none of them are found then return null; if found then return the node.

**Comment by rick-de-water** [comment]
Sure, getting it to work is not a problem. But with your example it would have to scan the entire class whenever there is any kind of change in a class, and I believe that the whole point of incremental generators is to do as little work as possible. I know that only changes to properties with attributes affect the code generation, so that is what I want to use as my starting point.

**Comment by bekir-ozturk** [comment]
I have ran into this as well, today.
My generator is creating static readonly byte arrays by encoding strings. Each string only needs to be encoded once since duplicates will be wasting memory. Running `Collect` to manually remove duplicates means that my generator won't be so _incremental_ as any addition or removal of a new string will end up regenerating the source for all the strings.

**Comment by CyrusNajmabadi** [comment]
@bekir-ozturk that is entirely fine.  If a new string is added, we'll run the next stage of the pipeline.  THat stage should collect things and produce the deduped list.  we'll then check that result agains the last computed version of that pipeline and not run further steps if was a duplicate.

**Comment by bekir-ozturk** [comment]
@CyrusNajmabadi  I think I get what you mean. I was passing my deduped list through the pipeline using an `IncrementalValueProvider<HashSet<string>>`, but now I converted it to an `IIncrementalValuesProvider<string>` by simply calling `SelectMany` on it.
Thanks!
